content/browser/ml/webnn/dml/readback_heap.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "content/browser/ml/webnn/dml/readback_heap.h"
+
+#include <memory>
+
+#include "content/browser/ml/webnn/dml/execution_context.h"
+
+namespace content::webnn {
+
+ReadbackHeap::ReadbackHeap(ExecutionContext* execution_context)
+    : execution_context_(execution_context), readback_resource_(nullptr) {}
+
+ReadbackHeap::~ReadbackHeap() = default;
+
+HRESULT ReadbackHeap::InitializeResource(std::vector<EdgeInfo>& outputs_info) {
+  uint64_t aligned_offset = 0;
+  for (auto& output_info : outputs_info) {
+    uint64_t byte_length = reinterpret_cast<const DML_BUFFER_TENSOR_DESC*>(
+                               output_info.outputTensorDESC.Desc)
+                               ->TotalTensorSizeInBytes;
+    MemoryInfo memory_info = {};
+    memory_info.byte_offset = aligned_offset;
+    memory_info.byte_length = byte_length;
+    outputs_info_map_[output_info.name] = memory_info;
+
+    // Only offset need to be algnement, the byte length keep original value.
+    aligned_offset += Align(byte_length, DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT);
+  }
+  outputs_resource_size_ = aligned_offset;
+  outputs_shm_region_ =
+      base::ReadOnlySharedMemoryRegion::Create(outputs_resource_size_);
+
+  HRESULT hr = CreateReadbackResource(outputs_resource_size_);
+  if (FAILED(hr)) {
+    return hr;
+  }
+  return S_OK;
+}
+
+// Readback inference result from GPU that is stored in named_outputs.
+HRESULT ReadbackHeap::ReadbackResource(NamedOutputsPtr& named_outputs,
+                                       ID3D12Resource* src_resource) {
+  // TODO:: Don't need add barrier to reset source resource from COPY_SOURCE
```

Comment:
Yes, you need to transition the output buffer back to UAV after this copy.

Reply:
Remove the TODO, thanks.